### PR TITLE
[2.3.2] Support Istio Gateway

### PR DESCRIPTION
**Pull request by loganhz**
Proposed changes
======
Support Istio Gateway

Types of changes
======
- New feature (non-breaking change which adds functionality)

Linked Issues
======
https://github.com/rancher/rancher/issues/23302
 

Further comments
======

### CODE REVIEW

**Comment by loganhz** [comment]
Backend merged

File: app/models/destinationrule.js [hunk 1]
```diff
@@ -20,11 +20,11 @@ export default Resource.extend({
 
   actions:      {
     edit() {
-      get(this, 'router').transitionTo('authenticated.project.istio.project-istio.destination-rules.detail.edit', get(this, 'id'));
+      get(this, 'router').transitionTo('authenticated.project.istio.destination-rule.detail.edit', get(this, 'id'));
```
Comment (westlywright) [inline app/models/destinationrule.js]: `this.router` && `this.id`

Comment (westlywright) [reply app/models/destinationrule.js]: Below as well.

File: app/models/gateway.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import Resource from '@rancher/ember-api-store/models/resource';
+import { reference } from '@rancher/ember-api-store/utils/denormalize';
+import { inject as service } from '@ember/service';
+import { get, computed } from '@ember/object';
+
```
Comment (westlywright) [inline app/models/gateway.js]: Use native getter's (`this.<param>`) in this file.

File: app/models/virtualservice.js [hunk 3]
```diff
@@ -16,11 +16,11 @@ export default Resource.extend({
 
   actions:      {
     edit() {
-      get(this, 'router').transitionTo('authenticated.project.istio.project-istio.virtual-services.detail.edit', get(this, 'id'));
+      get(this, 'router').transitionTo('authenticated.project.istio.virtual-service.detail.edit', get(this, 'id'));
```
Comment (westlywright) [inline app/models/virtualservice.js]: Use native getter's `(this.<param>)` in this file.

File: lib/istio/addon/components/cru-gateway/component.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  didInsertElement() {
+    if ( get(this, 'isNew') && get(this, 'servers.length') === 0 ) {
+      this.send('addServer');
+    }
+  },
+
+  actions: {
+    addServer() {
+      const server = EmberObject.create({
```
Comment (westlywright) [inline lib/istio/addon/components/cru-gateway/component.js]: Is there a reason this needs to be an EmberObject?

File: lib/istio/addon/components/cru-gateway/component.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import Component from '@ember/component';
+import ViewNewEdit from 'shared/mixins/view-new-edit';
+import layout from './template';
+import Errors from 'ui/utils/errors';
+import EmberObject from '@ember/object';
+import ChildHook from 'shared/mixins/child-hook';
+import { flattenLabelArrays } from 'shared/mixins/manage-labels';
+export default Component.extend(ViewNewEdit, ChildHook, {
+  intl: service(),
+
```
Comment (westlywright) [inline lib/istio/addon/components/cru-gateway/component.js]: Use native getter's `(this.<param>)` in this file.

File: lib/istio/addon/components/cru-gateway/component.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    set(this, 'errors', errors);
+
+    if ( get(errors, 'length') !== 0 ) {
+      return false;
+    }
+
+    return this.applyHooks('_beforeSaveHooks').then(() => {
+      set(this, 'model.namespaceId', get(this, 'namespace.id'));
+
+      return sup.apply(self, ...arguments);
```
Comment (westlywright) [inline lib/istio/addon/components/cru-gateway/component.js]: You're in an arrow function, can you use `this` here?

File: lib/istio/addon/components/cru-gateway/component.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }).catch((err) => {
+      set(this, 'errors', [Errors.stringify(err)]);
+    });
+  },
+
+  doSave(opt) {
+    const primaryResource = get(this, 'primaryResource');
+    let clone = primaryResource.clone();
```
Comment (westlywright) [inline lib/istio/addon/components/cru-gateway/component.js]: this.primaryResource.clone();

File: lib/istio/addon/components/cru-gateway/component.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        delete server['tls'];
+      }
+    });
+
+    opt = opt || {};
+    opt.qp = { '_replace': 'true' };
+
+    return clone.save(opt).then((newData) => {
+      return this.mergeResult(newData);
+    });
```
Comment (westlywright) [inline lib/istio/addon/components/cru-gateway/component.js]: error handled somewhere else?

File: lib/istio/addon/components/cru-virtual-service/component.js [hunk 9]
```diff
@@ -133,6 +129,14 @@ export default Component.extend(ViewNewEdit, ChildHook, {
     }
 
     (get(clone, 'http') || []).forEach((route) => {
+      (get(route, 'route') || []).forEach((destination) => {
```
Comment (westlywright) [inline lib/istio/addon/components/cru-virtual-service/component.js]: Use native getter's `(this.<param>)` in this file.

File: lib/istio/addon/components/gateway-server/template.hbs [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        {{/input-or-display}}
+      {{/form-value-array}}
+    </div>
+  </div>
+
+  {{#advanced-section advanced=advanced}}
+    <hr class="mt-20"/>
+    <div class="row">
+      <div class="col span-4">
+        <label class="acc-label">
```
Comment (westlywright) [inline lib/istio/addon/components/gateway-server/template.hbs]: Do we have issues with using the for/id semantics with `input-or-display`? If not i'd like to add the accessibility here and the other places we can in this file.

File: lib/istio/addon/components/istio-gateway-select/component.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+  namespaceId:  null,
+  value:        null,
+  showLabel:    true,
+
+  gateways:     null,
+
+  init() {
+    this._super(...arguments);
+    set(this, 'gateways', get(this, 'store').all('gateway'));
```
Comment (westlywright) [inline lib/istio/addon/components/istio-gateway-select/component.js]: Use native getter's `(this.<param>)` in this file.

File: lib/istio/addon/gateway/detail/edit/route.js [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import Route from '@ember/routing/route';
+import { hash } from 'rsvp';
+
+export default Route.extend({
+  model() {
+    const original = this.modelFor('gateway.detail').gateway;
```
Comment (westlywright) [inline lib/istio/addon/gateway/detail/edit/route.js]: If `gateway.detail` is empty we'd get an undefined error here?

File: lib/istio/addon/gateway/detail/edit/route.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+
+    return hash({ gateway: original.clone(),  });
```
Comment (westlywright) [inline lib/istio/addon/gateway/detail/edit/route.js]: I dont think you need hash here as you're not doing any promises. Its fine to return the object without the hash.

File: lib/istio/addon/project-istio/gateways/index/route.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import { get, setProperties } from '@ember/object';
+import { inject as service } from '@ember/service';
+import Route from '@ember/routing/route';
+import C from 'ui/utils/constants';
+
+export default Route.extend({
+  session:      service(),
+
+  model(params, transition) {
+    const store = get(this, 'store');
```
Comment (westlywright) [inline lib/istio/addon/project-istio/gateways/index/route.js]: Use native getter's `(this.<param>)` in this file.

**Pull request merged**